core/trino-main/src/main/java/io/trino/server/security/oauth2/IdpClientProvider.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.server.security.oauth2;
+
+import com.nimbusds.oauth2.sdk.AccessTokenResponse;
+import com.nimbusds.oauth2.sdk.AuthorizationCode;
+import com.nimbusds.oauth2.sdk.AuthorizationCodeGrant;
+import com.nimbusds.oauth2.sdk.AuthorizationRequest;
+import com.nimbusds.oauth2.sdk.ResponseType;
+import com.nimbusds.oauth2.sdk.Scope;
+import com.nimbusds.oauth2.sdk.TokenRequest;
+import com.nimbusds.oauth2.sdk.auth.ClientSecretBasic;
+import com.nimbusds.oauth2.sdk.auth.Secret;
+import com.nimbusds.oauth2.sdk.id.ClientID;
+import com.nimbusds.oauth2.sdk.id.State;
+import com.nimbusds.oauth2.sdk.token.AccessToken;
+import com.nimbusds.oauth2.sdk.token.Tokens;
+import com.nimbusds.openid.connect.sdk.AuthenticationRequest;
+import com.nimbusds.openid.connect.sdk.Nonce;
+import com.nimbusds.openid.connect.sdk.OIDCTokenResponse;
+import com.nimbusds.openid.connect.sdk.token.OIDCTokens;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+
+import java.net.URI;
+import java.time.Instant;
+import java.util.Optional;
+
+import static com.nimbusds.oauth2.sdk.ResponseType.Value.CODE;
+import static com.nimbusds.openid.connect.sdk.OIDCScopeValue.OPENID;
+import static java.util.Objects.requireNonNull;
+
+public class IdpClientProvider
+        implements Provider<IdpClient>
+{
+    private final ClientID clientId;
+    private final ClientSecretBasic clientAuth;
+    private final Scope scope;
+    private final URI authUrl;
+    private final URI tokenUrl;
+    private final NimbusHttpClient httpClient;
+
+    @Inject
+    public IdpClientProvider(OAuth2Config config, OAuth2Endpoints endpoints, NimbusHttpClient httpClient)
+    {
+        this.clientId = new ClientID(requireNonNull(config.getClientId(), "clientId is null"));
+        this.clientAuth = new ClientSecretBasic(clientId, new Secret(requireNonNull(config.getClientSecret(), "clientSecret is null")));
+        this.scope = Scope.parse(requireNonNull(config.getScopes(), "scopes is null"));
+        this.authUrl = URI.create(requireNonNull(endpoints.getAuthUrl(), "authUrl is null"));
+        this.tokenUrl = URI.create(requireNonNull(endpoints.getTokenUrl(), "tokenUrl is null"));
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+    }
+
+    @Override
+    public IdpClient get()
+    {
+        return scope.contains(OPENID) ? new OIDCStrategy() : new IdpStrategy();
```

Comment:
This can be decided at construction time, there's no need to wait until this method is called.